Tests/ConduitTests/Networking/Data/DataDownloaderTests.swift

```diff
@@ -0,0 +1,179 @@
+//
+//  DataDownloaderTests.swift
+//  ConduitTests
+//
+//  Created by Anthony Lipscomb on 8/3/21.
+//  Copyright © 2021 MINDBODY. All rights reserved.
+//
+
+import XCTest
+import Conduit
+
+private class MonitoringURLSessionClient: URLSessionClientType {
+    private let sessionClient = URLSessionClient()
+    var requestMiddleware: [RequestPipelineMiddleware] = []
+    var responseMiddleware: [ResponsePipelineMiddleware] = []
+
+    var numRequestsSent: Int = 0
+
+    func begin(request: URLRequest) throws -> (data: Data?, response: HTTPURLResponse) {
+        numRequestsSent += 1
+        return try sessionClient.begin(request: request)
+    }
+
+    func begin(request: URLRequest, completion: @escaping SessionTaskCompletion) -> SessionTaskProxyType {
+        numRequestsSent += 1
+        return sessionClient.begin(request: request, completion: completion)
+    }
+}
+
+class DataDownloaderTests: XCTestCase {
+
+    func testOnlyHitsNetworkOncePerRequest() throws {
+        let monitoringSessionClient = MonitoringURLSessionClient()
+        let sut = DataDownloader(cache: AutoPurgingURLDataCache(), sessionClient: monitoringSessionClient)
+        let url = try URL(absoluteString: "https://httpbin.org/image/svg")
+        let dataRequest = URLRequest(url: url)
+        for _ in 0..<100 {
+            sut.downloadData(for: dataRequest) { _ in }
+        }
+        XCTAssert(monitoringSessionClient.numRequestsSent == 1)
+    }
+
+    func testMarksDatasAsCachedAfterDownloaded() throws {
+        let attemptedAllDataRetrievalsExpectation = expectation(description: "attempted all data retrievals")
+
+        let sut = DataDownloader(cache: AutoPurgingURLDataCache())
+        let url = try URL(absoluteString: "https://httpbin.org/image/svg")
+        let dataRequest = URLRequest(url: url)
+
+        sut.downloadData(for: dataRequest) { response in
+            XCTAssert(response.isFromCache == false)
+
+            let dispatchGroup = DispatchGroup()
+
+            for _ in 0..<100 {
+                dispatchGroup.enter()
+                DispatchQueue.global().async {
+                    sut.downloadData(for: dataRequest) { response in
+                        XCTAssert(response.isFromCache == true)
+                        dispatchGroup.leave()
+                    }
+                }
+            }
+            dispatchGroup.notify(queue: DispatchQueue.main) {
+                attemptedAllDataRetrievalsExpectation.fulfill()
+            }
```

Comment:
I'm thinking it might be better to set the expectation fulfillment count here:
```
attemptedAllDataRetrievalsExpectation.expectedFulfillmentCount = 100
```

and then with each downloadData completion have 
```
attemptedAllDataRetrievalsExpectation.fulfill()
```

and then there is no need for this `dispatchGroup.notify`